### PR TITLE
[wallet] use status-go wallet service

### LINKED ISSUE #1762: Wallet - Use web3 to fetch wallet transactions
[comment]: # (Please replace ... with your information. Remove < and >)

### User Story
As a developper, I want to to use web3 to fetch transactions so that Status does not depends on a centralized api (etherscan) for wallet transactions history.

### Description
[comment]: # (Feature or Bug? i.e Type: Bug)
_Type_: Feature

[comment]: # (Describe the feature you would like, or briefly summarise the bug and what you did, what you expected to happen, and what actually happens. Sections below)
_Summary_: 
To reach MVP of the new wallet we used etherscan to fetch the transaction history of the wallet. This task will consist of implementing a new version of the transaction fetching logic using web3.

#### Expected behavior
[comment]: # (Describe what you expected to happen.)
Transactions are fetched using web3

#### Actual behavior
[comment]: # (Describe what actually happened.)
Transactions are fetched using etherscan and are limited to the 20 last transactions

### Solution
[comment]: # (Please summarise the solution and provide a task list on what needs to be fixed.)
_Summary_: 

* [ ] implement and adapt this script in clojurescript https://ethereum.stackexchange.com/questions/2531/common-useful-javascript-snippets-for-geth/3478#3478 to get transactions for a given address from block X to Z in utils/transactions.cljs
* [ ] create an event that recursively calls the function for block n to n-step until the block when the wallet has been created is reached in ui.screens.wallet.events

Depending on the performance cost of this solution it might be worth it to keep the etherscan api call for the 20 first transactions, but use web3 as fallback in case it fails and fetch the rest of the transactions with web3.

### PR DESCRIPTION
**Pull request by yenda**
Fixes #1762

### Summary

- remove use of etherscan and subscriptions
- replace by status-go

#### Areas that maybe impacted
Wallet transaction history

Tested in emulator for  `satoshi document engage inflict goddess auction rule unfair bid next buddy shy` on ropsten

### Steps to test

status: ready

### CODE REVIEW

**Comment by yenda** [comment]
When recovering an account (probably same with login I get this:

```
5  4885  5113 D StatusModule: Starting Geth node in folder: /data/user/0/im.status.ethereum.debug/ethereum/testnet
07-09 10:21:51.948  4885  5113 D StatusModule: startNode() with config (see below)
07-09 10:21:51.948  4885  5113 D StatusModule: ********************** NODE CONFIG ****************************
07-09 10:21:51.948  4885  5113 D StatusModule: Node config:{"ClusterConfig":{"Enabled":true,"Fleet":"eth.beta","BootNodes":["enode:\/\/5395aab7833f1ecb671b59bf0521cf20224fe8162fc3d2675de4ee4d5636a75ec32d13268fc184df8d1ddfa803943906882da62a4df42d4fccf6d17808156a87@206.189.243.57:443","enode:\/\/7427dfe38bd4cf7c58bb96417806fab25782ec3e6046a8053370022cbaa281536e8d64ecd1b02e1f8f72768e295d06258ba43d88304db068e6f2417ae8bcb9a6@104.154.88.123:443","enode:\/\/ebefab39b69bbbe64d8cd86be765b3be356d8c4b24660f65d493143a0c44f38c85a257300178f7845592a1b0332811542e9a58281c835babdd7535babb64efc1@35.202.99.224:443","enode:\/\/436cc6f674928fdc9a9f7990f2944002b685d1c37f025c1be425185b5b1f0900feaf1ccc2a6130268f9901be4a7d252f37302c8335a2c1a62736e9232691cc3a@174.138.105.243:443"],"TrustedMailServers":["enode:\/\/8a64b3c349a2e0ef4a32ea49609ed6eb3364be1110253c20adc17a3cebbc39a219e5d3e13b151c0eee5d8e0f9a8ba2cd026014e67b41a4ab7d1d5dd67ca27427@206.189.243.168:443","enode:\/\/531e252ec966b7e83f5538c19bf1cde7381cc7949026a6e499b6e998e695751aadf26d4c98d5a4eabfb7cefd31c3c88d600a775f14ed5781520a88ecd25da3c6@35.225.227.79:443","enode:\/\/da61e9eff86a56633b635f887d8b91e0ff5236bbc05b8169834292e92afb92929dcf6efdbf373a37903da8fe0384d5a0a8247e83f1ce211aa429200b6d28c548@47.91.156.93:443","enode:\/\/015e22f6cd2b44c8a51bd7a23555e271e0759c7d7f52432719665a74966f2da456d28e154e836bee6092b4d686fe67e331655586c57b718be3997c1629d24167@35.226.21.19:443","enode:\/\/7aa648d6e855950b2e3d3bf220c496e0cae4adfddef3e1e6062e6b177aec93bc6cdcf1282cb40d1656932ebfdd565729da440368d7c4da7dbd4d004b1ac02bf8@206.189.243.169:443","enode:\/\/7de99e4cb1b3523bd26ca212369540646607c721ad4f3e5c821ed9148150ce6ce2e72631723002210fac1fd52dfa8bbdf3555e05379af79515e1179da37cc3db@35.188.19.210:443"],"StaticNodes":["enode:\/\/887cbd92d95afc2c5f1e227356314a53d3d18855880ac0509e0c0870362aee03939d4074e6ad31365915af41d34320b5094bfcc12a67c381788cd7298d06c875@206.189.243.177:443","enode:\/\/aa97756bc147d74be6d07adfc465266e17756339d3d18591f4be9d1b2e80b86baf314aed79adbe8142bcb42bc7bc40e83ee3bbd0b82548e595bf855d548906a1@47.52.188.241:443"],"RendezvousNodes":["\/ip4\/174.138.105.243\/tcp\/30703\/ethv4\/16Uiu2HAmRHPzF3rQg55PgYPcQkyvPVH9n2hWsYPhUJBZ6kVjJgdV","\/ip4\/206.189.243.57\/tcp\/30703\/ethv4\/16Uiu2HAmLqTXuY4Sb6G28HNooaFUXUKzpzKXCcgyJxgaEE2i5vnf"]},"DataDir":"\/data\/user\/0\/im.status.ethereum.debug\/ethereum\/mainnet_rpc_dev","LogLevel":"INFO","Rendezvous":true,"WhisperConfig":{"Enabled":true,"LightClient":true,"MinimumPoW":0.001,"EnableNTPSync":true},"LogEnabled":true,"RequireTopics":{"whisper":{"Min":2,"Max":2}},"UpstreamConfig":{"Enabled":true,"URL":"https:\/\/mainnet.infura.io\/v3\/f315575765b14720b32382a61a89341a"},"ListenAddr":":30304","NetworkId":1,"Name":"StatusIM","NoDiscovery":false,"ShhextConfig":{"BackupDisabledDataDir":"\/data\/user\/0\/im.status.ethereum.debug\/files\/..\/no_backup","InstallationID":"94b2a168-291f-59f7-a962-d29f5cc9acfb","MaxMessageDeliveryAttempts":6,"MailServerConfirmations":true,"PFSEnabled":true},"WalletConfig":{"Enabled":true},"KeyStoreDir":"\/data\/user\/0\/im.status.ethereum.debug\/keystore"}
07-09 10:21:51.948  4885  5113 D StatusModule: ******************* ENDOF NODE CONFIG *************************
07-09 10:21:51.958  4885  5113 D StatusModule: StartNode result: {"error":""}
07-09 10:21:51.958  4885  5113 D StatusModule: Geth node started
07-09 10:21:51.987  4885  4954 E GoLog   : INFO [07-09|08:21:51.986] LES protocol is disabled                 package=status-go/node
07-09 10:21:51.987  4885  4957 E GoLog   : INFO [07-09|08:21:51.987] Incentivisation is disabled              package=status-go/node
07-09 10:21:51.987  4885  4957 E GoLog   : INFO [07-09|08:21:51.987] Status service api is disabled           package=status-go/node
07-09 10:21:51.989  4885  4957 E GoLog   : INFO [07-09|08:21:51.989] Starting peer-to-peer node               instance=StatusIM/vwallet-feedback/android-386/go1.11.5
07-09 10:21:52.014  4885  4954 E GoLog   : INFO [07-09|08:21:52.014] started whisper v.6.0 
07-09 10:21:52.021  4885  5114 E GoLog   : INFO [07-09|08:21:52.021] whisper stopped 
07-09 10:21:52.024  4885  5114 E GoLog   : INFO [07-09|08:21:52.024] New local node record                    seq=1 id=c482bb99b1b8a489 ip=127.0.0.1 udp=0 tcp=30304
07-09 10:21:52.024  4885  5114 E GoLog   : INFO [07-09|08:21:52.024] Started P2P networking                   self="enode://4d8109e60fa07597986137d9a42a2c0a68bf37110aec4ad6c79a4faeaeb9e828bccf898d24ed400be7b655dd434bec85bb310831102d4d7fc6687aeedb2c9505@127.0.0.1:30304?discport=0"
07-09 10:21:52.629  1982  2399 I GnssLocationProvider: WakeLock acquired by sendMessage(REPORT_SV_STATUS, 0, com.android.server.location.GnssLocationProvider$SvStatusInfo@cc3e4a1)
07-09 10:21:52.630  1982  2000 I GnssLocationProvider: WakeLock released by handleMessage(REPORT_SV_STATUS, 0, com.android.server.location.GnssLocationProvider$SvStatusInfo@cc3e4a1)
07-09 10:21:53.629  1982  2399 I GnssLocationProvider: WakeLock acquired by sendMessage(REPORT_SV_STATUS, 0, com.android.server.location.GnssLocationProvider$SvStatusInfo@49796c6)
07-09 10:21:53.631  1982  2000 I GnssLocationProvider: WakeLock released by handleMessage(REPORT_SV_STATUS, 0, com.android.server.location.GnssLocationProvider$SvStatusInfo@49796c6)
07-09 10:21:54.126  4885  4951 D StatusModule: Signal event: {"type":"node.crashed","event":{"error":"protocol is not initialized"}}
07-09 10:21:54.146  4885  4947 D ReactNativeJS: DEBUG [status-im.utils.handlers:32] - Handling re-frame event:  :signals/signal-received
07-09 10:21:54.165  4885  4947 D ReactNativeJS: DEBUG [status-im.events:862] - :event-str {"type":"node.crashed","event":{"error":"protocol is not initialized"}}
07-09 10:21:54.194  4885  4947 D ReactNativeJS: DEBUG [status-im.signals.core:84] - Event  node.crashed  not handled {:error "protocol is not initialized"}
07-09 10:21:54.629  1982  2399 I GnssLocationProvider: WakeLock acquired by sendMessage(REPORT_SV_STATUS, 0, com.android.server.location.GnssLocationProvider$SvStatusInfo@79fb687)
07-09 10:21:54.630  1982  2000 I GnssLocationProvider: WakeLock released by handleMessage(REPORT_SV_STATUS, 0, com.android.server.location.GnssLocationProvider$SvStatusInfo@79fb687)
07-09 10:21:54.993  1640  1721 W audio_hw_generic: Not supplying enough data to HAL, expected position 2141375 , only wrote 1797840
07-09 10:21:55.630  1982  2399 I GnssLocationProvider: WakeLock acquired by sendMessage(REPORT_SV_STATUS, 0, com.android.server.location.GnssLocationProvider$SvStatusInfo@c0809b4)
07-09 10:21:55.631  1982  2000 I GnssLocationProvider: WakeLock released by handleMessage(REPORT_SV_STATUS, 0, com.android.server.location.GnssLocationProvider$SvStatusInfo@c0809b4)
07-09 10:21:56.629  1982  2399 I GnssLocationProvider: WakeLock acquired by sendMessage(REPORT_SV_STATUS, 0, com.android.server.location.GnssLocationProvider$SvStatusInfo@79929dd)
07-09 10:21:56.630  1982  2000 I GnssLocationProvider: WakeLock released by handleMessage(REPORT_SV_STATUS, 0, com.android.server.location.GnssLocationProvider$SvStatusInfo@79929dd)
07-09 10:21:57.629  1982  2399 I GnssLocationProvider: WakeLock acquired by sendMessage(REPORT_SV_STATUS, 0, com.android.server.location.GnssLocationProvider$SvStatusInfo@4e65952)
07-09 10:21:57.630  1982  2000 I GnssLocationProvider: WakeLock released by handleMessage(REPORT_SV_STATUS, 0, com.android.server.location.GnssLocationProvider$SvStatusInfo@4e65952)
07-09 10:21:58.629  1982  2399 I GnssLocationProvider: WakeLock acquired by sendMessage(REPORT_SV_STATUS, 0, com.android.server.location.GnssLocationProvider$SvStatusInfo@cb73423)
07-09 10:21:58.630  1982  2000 I GnssLocationProvider: WakeLock released by handleMessage(REPORT_SV_STATUS, 0, com.android.server.location.GnssLocationProvider$SvStatusInfo@cb73423)
07-09 10:21:59.629  1982  2399 I GnssLocationProvider: WakeLock acquired by sendMessage(REPORT_SV_STATUS, 0, com.android.server.location.GnssLocationProvider$SvStatusInfo@1f60d20)
07-09 10:21:59.630  1982  2000 I GnssLocationProvider: WakeLock released by handleMessage(REPORT_SV_STATUS, 0, com.android.server.location.GnssLocationProvider$SvStatusInfo@1f60d20)
07-09 10:22:00.004  1982  1982 I ActivityManager: Killing 2868:com.google.process.gapps/u0a11 (adj 906): empty #17
07-09 10:22:00.011  1982  2004 W libprocessgroup: kill(-2868, 9) failed: No such process
07-09 10:22:00.095  1982  2004 I chatty  : uid=1000(system) ActivityManager identical 2 lines
07-09 10:22:00.135  1982  2004 W libprocessgroup: kill(-2868, 9) failed: No such process
07-09 10:22:00.143  1804  1804 I Zygote  : Process 2868 exited due to signal (9)
07-09 10:22:00.175  1982  2004 W libprocessgroup: kill(-2868, 9) failed: No such process
07-09 10:22:00.175  1982  2004 I libprocessgroup: Successfully killed process cgroup uid 10011 pid 2868 in 165ms
07-09 10:22:00.629  1982  2399 I GnssLocationProvider: WakeLock acquired by sendMessage(REPORT_SV_STATUS, 0, com.android.server.location.GnssLocationProvider$SvStatusInfo@3ecb89e)
07-09 10:22:00.630  1982  2000 I GnssLocationProvider: WakeLock released by handleMessage(REPORT_SV_STATUS, 0, com.android.server.location.GnssLocationProvider$SvStatusInfo@3ecb89e)
07-09 10:22:01.629  1982  2399 I GnssLocationProvider: WakeLock acquired by sendMessage(REPORT_SV_STATUS, 0, com.android.server.location.GnssLocationProvider$SvStatusInfo@43eef7f)
07-09 10:22:01.630  1982  2000 I GnssLocationProvider: WakeLock released by handleMessage(REPORT_SV_STATUS, 0, com.android.server.location.GnssLocationProvider$SvStatusInfo@43eef7f)
07-09 10:22:02.629  1982  2399 I GnssLocationProvider: WakeLock acquired by sendMessage(REPORT_SV_STATUS, 0, com.android.server.location.GnssLocationProvider$SvStatusInfo@6fedb4c)
07-09 10:22:02.630  1982  2000 I GnssLocationProvider: WakeLock released by handleMessage(REPORT_SV_STATUS, 0, com.android.server.location.GnssLocationProvider$SvStatusInfo@6fedb4c)
07-09 10:22:03.630  1982  2399 I GnssLocationProvider: WakeLock acquired by sendMessage(REPORT_SV_STATUS, 0, com.android.server.location.GnssLocationProvider$SvStatusInfo@a1cb795)
07-09 10:22:03.632  1982  2000 I GnssLocationProvider: WakeLock released by handleMessage(REPORT_SV_STATUS, 0, com.android.server.location.GnssLocationProvider$SvStatusInfo@a1cb795)
07-09 10:22:04.629  1982  2399 I GnssLocationProvider: WakeLock acquired by sendMessage(REPORT_SV_STATUS, 0, com.android.server.location.GnssLocationProvider$SvStatusInfo@67c0aa)
07-09 10:22:04.630  1982  2000 I GnssLocationProvider: WakeLock released by handleMessage(REPORT_SV_STATUS, 0, com.android.server.location.GnssLocationProvider$SvStatusInfo@67c0aa)
07-09 10:22:05.630  1982  2399 I GnssLocationProvider: WakeLock acquired by sendMessage(REPORT_SV_STATUS, 0, com.android.server.location.GnssLocationProvider$SvStatusInfo@65d049b)
07-09 10:22:05.631  1982  2000 I GnssLocationProvider: WakeLock released by handleMessage(REPORT_SV_STATUS, 0, com.android.server.location.GnssLocationProvider$SvStatusInfo@65d049b)
07-09 10:22:06.629  1982  2399 I GnssLocationProvider: WakeLock acquired by sendMessage(REPORT_SV_STATUS, 0, com.android.server.location.GnssLocationProvider$SvStatusInfo@7e46038)
07-09 10:22:06.631  1982  2000 I GnssLocationProvider: WakeLock released by handleMessage(REPORT_SV_STATUS, 0, com.android.server.location.GnssLocationProvider$SvStatusInfo@7e46038)
07-09 10:22:07.629  1982  2399 I GnssLocationProvider: WakeLock acquired by sendMessage(REPORT_SV_STATUS, 0, com.android.server.location.GnssLocationProvider$SvStatusInfo@4de3811)
07-09 10:22:07.629  1982  2000 I GnssLocationProvider: WakeLock released by handleMessage(REPORT_SV_STATUS, 0, com.android.server.location.GnssLocationProvider$SvStatusInfo@4de3811)
07-09 10:22:08.145  1918  1918 E netmgr  : Failed to open QEMU pipe 'qemud:network': Invalid argument
07-09 10:22:08.145  1918  1918 E netmgr  : WifiForwarder unable to open QEMU pipe: Invalid argument
07-09 10:22:08.629  1982  2399 I GnssLocationProvider: WakeLock acquired by sendMessage(REPORT_SV_STATUS, 0, com.android.server.location.GnssLocationProvider$SvStatusInfo@5d93d76)
07-09 10:22:08.630  1982  2000 I GnssLocationProvider: WakeLock released by handleMessage(REPORT_SV_STATUS, 0, com.android.server.location.GnssLocationProvider$SvStatusInfo@5d93d76)
07-09 10:22:09.629  1982  2399 I GnssLocationProvider: WakeLock acquired by sendMessage(REPORT_SV_STATUS, 0, com.android.server.location.GnssLocationProvider$SvStatusInfo@d7d4f77)
07-09 10:22:09.629  1982  2000 I GnssLocationProvider: WakeLock released by handleMessage(REPORT_SV_STATUS, 0, com.android.server.location.GnssLocationProvider$SvStatusInfo@d7d4f77)
07-09 10:22:10.418  1982  2047 I ActivityManager: Killing 4236:com.google.android.videos/u0a66 ```

**Comment by yenda** [comment]
@cammellos @dshulyak depends on the release of status-go wallet-feedback branch

**Comment by cammellos** [comment]
@yenda fix has been merged, you can use top of develop

**Comment by cammellos** [comment]
v0.30.0-beta.1

**Comment by yenda** [comment]
@cammellos it is this one I need https://github.com/status-im/status-go/pull/1517/files

**Comment by cammellos** [comment]
@yenda v0.30.0-beta.1 should include all changes + the fix, are you having some issues with it?

**Comment by yenda** [comment]
@cammellos it doesn't include the PR wallet-feedback which just got merged

**Comment by cammellos** [comment]
@yenda ah, sorry I didn't notice it was just merged

File: src/status_im/ethereum/encode.cljs [hunk 1]
```diff
@@ -0,0 +1,8 @@
+(ns status-im.ethereum.encode
+  (:require [status-im.js-dependencies :as dependencies]))
+
+(defn utils [] (dependencies/web3-utils))
+
+(defn uint
+  [x]
+  (.numberToHex (utils) x))
```
Comment (vitvly) [inline src/status_im/ethereum/encode.cljs:8]: Will memoize for `utils` have any impact here?

**Review by vitvly — APPROVED** [review]
Nitpick: Some tests would be nice for new fns.

File: src/status_im/ethereum/subscriptions.cljs [hunk 2]
```diff
@@ -22,82 +23,37 @@
...
+              (cond-> {}
+                (not historical?)
+                (assoc :db (assoc db [:ethereum/current-block] block-number))
+
+                (not-empty accounts)
+                (assoc ::transactions/get-transfers {:chain-tokens chain-tokens
+                                                     :from-block block-number}))
+              (transactions/check-watched-transactions))))
+
+(fx/defn reorg
```
Comment (flexsurfer) [inline src/status_im/ethereum/subscriptions.cljs:42]: what's reorg means?

File: src/status_im/ethereum/transactions/core.cljs [hunk 3]
```diff
@@ -2,14 +2,27 @@
...
+(def etherscan-supported? #{:testnet :mainnet :rinkeby})
+
+(let [network->subdomain {:testnet "ropsten" :rinkeby "rinkeby"}]
+  (defn get-transaction-details-url [chain hash]
+    {:pre [(keyword? chain) (string? hash)]
+     :post [(or (nil? %) (string? %))]}
+    (when (etherscan-supported? chain)
+      (let [network-subdomain (when-let [subdomain (network->subdomain chain)]
+                                (str subdomain "."))]
+        (str "https://" network-subdomain "etherscan.io/tx/" hash)))))
```
Comment (flexsurfer) [inline src/status_im/ethereum/transactions/core.cljs:24]: we still need etherescan? we can't abandon it?

**Review by flexsurfer — APPROVED** [review]

File: src/status_im/ethereum/transactions/core.cljs (hunk 3 above)
Comment (yenda) [reply src/status_im/ethereum/transactions/core.cljs:24]: this is just for showing the transaction details URL if someone wants to check it out in the browser, it's not etherscan api

Comment (flexsurfer) [reply src/status_im/ethereum/transactions/core.cljs:24]: right! thanks

File: src/status_im/ethereum/subscriptions.cljs (hunk 2 above)
Comment (yenda) [reply src/status_im/ethereum/subscriptions.cljs:42]: it's block reorg, sometime you receive blocks only to later learn that they were wrong and you have to reorg your chain. what happens here is that status-go tells you "they fucked up drop everything down to that block" and then you ask for historical transfers from that point to now to recover

File: src/status_im/ethereum/encode.cljs (hunk 1 above)
Comment (yenda) [reply src/status_im/ethereum/encode.cljs:8]: hard to tell I think it is ok for now, also we are discussing with @dshulyak the idea of receiving transit encoded data and status-go would do the conversion, but I tested on an account with 200 transfers and it doesn't seem to be an issue

**Comment by asemiankevich** [comment]
@yenda 

when i recover account in this build i see errors:

image


image

No special steps, just recover account and wait for a while.

I cant reproduce in nightly. Could this be related to your branch?

**Comment by Serhy** [comment]
Few issues from me:

1) I was expecting that moving transaction history to status-go level would speed up list of transactions appearance on History tab, but it seems not. With the account having long history (https://etherscan.io/address/0x8443b668be105143466BCDb8d8Fce1f697762eD3) it took ~1 minute to start displaying transactions (00:25 - 01:23 in video below)

2) When fetching of transaction history starts it does it from the beginning of when Wallet created. Is it possible to do it in reverse order, so that very latest transactions are appearing first till block when the wallet has been created? On the video below (01:21 - 02:30) we could see populating of tx history.

3) On the Transaction details view - the Confirmation value is 0 on any TX:
https://etherscan.io/tx/0x578cece512c62b07273b332033381e95c5c6ccf29aa8d64b05393b9c3d712a3d

Screenshot 2019-07-12 at 15 20 26

Video for the above issues: https://drive.google.com/file/d/1SATRmr0DK7sKcr--sCTa5fIr-oE4bIDu/view?usp=sharing - Galaxy Note 4 with Android 6.0.1 is the device on video.

However, after all transaction history list populated - it's all fine and no "syncing tx history" problems (i.e. 1st and 2nd items are not relevant), though if we could fix 1st and 2nd items above would be great.

**Comment by yenda** [comment]
@Serhy 

1. can't be fixed, on a freshly imported account, it was faster before because the list is fetched from etherscan API, which we don't depend anymore now by moving that to status-go. It is only a slower the first time because status-go is caching these transactions
2. could be checked with @dshulyak if that is possible, probably
3. will look at it

**Comment by dshulyak** [comment]
> could be checked with @dshulyak if that is possible, probably

fetching transaction in forward order made implementation easier to understand, but i will look into this problem. maybe it can be improved

**Comment by yenda** [comment]
superseeded by https://github.com/status-im/status-react/pull/8552

**Pull request closed without merging**